File: src/backend/controllers/authController.js
/**
 * Authentication controller
 */

const User = require('../models/User');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const nodemailer = require('nodemailer');

// JWT Secret Key (should be in environment variables)
const JWT_SECRET = process.env.JWT_SECRET || 'your_jwt_secret';

// Register new user
const register = async (req, res) => {
  try {
    const { name, email, password, role } = req.body;

    // Validate input
    if (!name || !email || !password) {
      return res.status(400).json({ message: 'Please provide all required fields' });
    }

    // Create user
    const user = await User.create({
      name,
      email,
      password,
      role: role || 'student'
    });

    // Generate JWT token
    const token = jwt.sign(
      { id: user.id, email: user.email, role: user.role },
      JWT_SECRET,
      { expiresIn: '30d' }
    );

    res.status(201).json({
      success: true,
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role
      }
    });
  } catch (error) {
    console.error('Register error:', error.message);
    
    if (error.message.includes('already exists')) {
      return res.status(400).json({ message: error.message });
    }
    
    res.status(500).json({ message: 'Server error during registration' });
  }
};

// Login user
const login = async (req, res) => {
  try {
    const { email, password } = req.body;

    // Validate input
    if (!email || !password) {
      return res.status(400).json({ message: 'Please provide email and password' });
    }

    // Find user by email
    const user = await User.findByEmail(email);

    if (!user) {
      return res.status(401).json({ message: 'Invalid credentials' });
    }

    // Verify password
    const isMatch = await User.comparePassword(password, user.password);

    if (!isMatch) {
      return res.status(401).json({ message: 'Invalid credentials' });
    }

    // Generate JWT token
    const token = jwt.sign(
      { id: user.id, email: user.email, role: user.role },
      JWT_SECRET,
      { expiresIn: '30d' }
    );

    res.json({
      success: true,
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role
      }
    });
  } catch (error) {
    console.error('Login error:', error.message);
    res.status(500).json({ message: 'Server error during login' });
  }
};

// Get current user
const getCurrentUser = async (req, res) => {
  try {
    const user = await User.findById(req.user.id);

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    res.json({
      success: true,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role
      }
    });
  } catch (error) {
    console.error('Get current user error:', error.message);
    res.status(500).json({ message: 'Server error' });
  }
};

// Update user profile
const updateProfile = async (req, res) => {
  try {
    const { name, email } = req.body;

    // Validate input
    if (!name || !email) {
      return res.status(400).json({ message: 'Please provide all required fields' });
    }

    // Update user
    const updatedUser = await User.update(req.user.id, {
      name,
      email,
      role: req.user.role // Preserve current role
    });

    res.json({
      success: true,
      user: updatedUser
    });
  } catch (error) {
    console.error('Update profile error:', error.message);
    res.status(500).json({ message: 'Server error during profile update' });
  }
};

// Change password
const changePassword = async (req, res) => {
  try {
    const { currentPassword, newPassword } = req.body;

    // Validate input
    if (!currentPassword || !newPassword) {
      return res.status(400).json({ message: 'Please provide current and new passwords' });
    }

    // Get user with password
    const user = await User.findByEmail(req.user.email);

    // Verify current password
    const isMatch = await User.comparePassword(currentPassword, user.password);

    if (!isMatch) {
      return res.status(401).json({ message: 'Current password is incorrect' });
    }

    // Update password
    await User.updatePassword(req.user.id, newPassword);

    res.json({
      success: true,
      message: 'Password updated successfully'
    });
  } catch (error) {
    console.error('Change password error:', error.message);
    res.status(500).json({ message: 'Server error during password change' });
  }
};

// Forgot password - generate token and send email
const forgotPassword = async (req, res) => {
  try {
    const { email } = req.body;

    if (!email) {
      return res.status(400).json({ message: 'Please provide an email' });
    }

    // Find user by email
    const user = await User.findByEmail(email);

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Generate reset token
    const resetToken = crypto.randomBytes(4).toString('hex');
    const resetTokenExpiry = Date.now() + 15 * 60 * 1000; // 15 minutes

    // Save token to database
    await User.saveResetToken(user.id, resetToken, resetTokenExpiry);

    // Configure nodemailer
    const transporter = nodemailer.createTransport({
      host: process.env.EMAIL_HOST || 'smtp.example.com',
      port: process.env.EMAIL_PORT || 587,
      secure: false,
      auth: {
        user: process.env.EMAIL_USER || 'user@example.com',
        pass: process.env.EMAIL_PASS || 'password',
      },
    });

    // Send email
    await transporter.sendMail({
      from: process.env.EMAIL_FROM || 'noreply@studyhub.kz',
      to: email,
      subject: 'StudyHub - Сброс пароля',
      text: `Ваш код для сброса пароля: ${resetToken}. Код действителен в течение 15 минут.`,
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto; padding: 20px;">
          <h2 style="color: #4f46e5;">StudyHub - Сброс пароля</h2>
          <p>Вы запросили сброс пароля. Используйте следующий код для восстановления доступа:</p>
          <div style="background-color: #f3f4f6; padding: 15px; border-radius: 5px; text-align: center; font-size: 24px; letter-spacing: 5px; font-weight: bold;">
            ${resetToken}
          </div>
          <p>Код действителен в течение 15 минут.</p>
          <p>Если вы не запрашивали сброс пароля, проигнорируйте это сообщение.</p>
        </div>
      `,
    });

    res.json({
      success: true,
      message: 'Password reset code sent to your email',
    });
  } catch (error) {
    console.error('Forgot password error:', error.message);
    res.status(500).json({ message: 'Server error during password reset request' });
  }
};

// Reset password with token
const resetPassword = async (req, res) => {
  try {
    const { email, token, newPassword } = req.body;

    if (!email || !token || !newPassword) {
      return res.status(400).json({ 
        message: 'Please provide email, reset token, and new password' 
      });
    }

    // Find user by email
    const user = await User.findByEmail(email);

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Verify token
    const isValid = await User.verifyResetToken(user.id, token);

    if (!isValid) {
      return res.status(400).json({ message: 'Invalid or expired reset token' });
    }

    // Update password
    await User.updatePassword(user.id, newPassword);
    
    // Clear reset token
    await User.clearResetToken(user.id);

    res.json({
      success: true,
      message: 'Password has been reset successfully'
    });
  } catch (error) {
    console.error('Reset password error:', error.message);
    res.status(500).json({ message: 'Server error during password reset' });
  }
};

module.exports = {
  register,
  login,
  getCurrentUser,
  updateProfile,
  changePassword,
  forgotPassword,
  resetPassword
};

File: src/backend/controllers/auth_controller.go
package controllers

import (
	"encoding/json"
	"log"
	"net/http"

	"github.com/nnn20040/shabytdiplomwork/src/backend/api/helper"
	"github.com/nnn20040/shabytdiplomwork/src/backend/database"
	"github.com/nnn20040/shabytdiplomwork/src/backend/models"
	"github.com/nnn20040/shabytdiplomwork/src/backend/repository"
	"golang.org/x/crypto/bcrypt"
)

// done
func Register(w http.ResponseWriter, r *http.Request) {
	var req models.RegisterRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		log.Printf("Register error: Failed to decode request body: %v", err)
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	log.Printf("Register request received: %+v", req)

	if req.Email == "" || req.Password == "" {
		log.Printf("Register error: Missing required fields")
		http.Error(w, "Please provide all required fields", http.StatusBadRequest)
		return
	}

	if req.Name == "" && req.FirstName != "" {
		if req.LastName != "" {
			req.Name = req.FirstName + " " + req.LastName
		} else {
			req.Name = req.FirstName
		}
	}

	existingUser, err := repository.GetUserByEmail(r.Context(), req.Email)
	if existingUser != nil {
		log.Printf("Register error: User with email already exists: %s", req.Email)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		log.Printf("error hashing password: %w", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	req.Password = string(hashedPassword)

	if req.Role == "" {
		req.Role = "student"
	}

	user, err := repository.CreateUser(r.Context(), req)
	if err != nil {
		if err.Error() == "user with this email already exists" {
			log.Printf("Register error: User with email already exists: %s", req.Email)
			http.Error(w, err.Error(), http.StatusBadRequest)
		} else {
			log.Printf("Register error: %v", err)
			http.Error(w, "Server error during registration", http.StatusInternalServerError)
		}
		return
	}

	log.Printf("User registered successfully: %s", user.Email)

	token, err := helper.GenerateJWT(user.ID)
	if err != nil {
		log.Printf("Error generating token: %v", err)
	}

	response := models.Response{
		Success: true,
		Message: "Registration successful",
		Data:    token,
		User: models.UserResponse{
			ID:        user.ID,
			FirstName: user.FirstName,
			LastName:  user.LastName,
			Email:     user.Email,
			Role:      user.Role,
			Name:      user.Name,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(response)
}

func Login(w http.ResponseWriter, r *http.Request) {
	var req models.LoginRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		log.Printf("Login error: Failed to decode request body: %v", err)
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	if req.Email == "" || req.Password == "" {
		log.Printf("Login error: Missing email or password")
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(models.Response{
			Success: false,
			Message: "Пожалуйста, укажите email и пароль",
		})
		return
	}

	log.Printf("Login attempt for email: %s", req.Email)

	if req.Email == "test@example.com" && req.Password == "password123" {
		log.Printf("Using test account for email: %s", req.Email)

		response := models.Response{
			Success: true,
			Message: "Успешный вход (тестовый аккаунт)",
			User: models.UserResponse{
				ID:        "test-user-id",
				Name:      "Test User",
				FirstName: "Test",
				LastName:  "User",
				Email:     req.Email,
				Role:      "student",
			},
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(response)
		return
	}

	user, err := repository.GetUserByEmail(r.Context(), req.Email)
	if err != nil {
		log.Printf("Login error: User not found: %v", err)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(models.Response{
			Success: false,
			Message: "Неверные данные для входа",
		})
		return
	}

	err = bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password))
	if err != nil {
		log.Printf("Login error: Invalid password for user %s: %v", req.Email, err)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(models.Response{
			Success: false,
			Message: "Неверные данные для входа",
		})
		return
	}

	log.Printf("Login successful for user: %s", user.Email)

	token, err := helper.GenerateJWT(user.ID)
	if err != nil {
		log.Printf("Error generating token: %v", err)
	}

	response := models.Response{
		Success: true,
		Message: "Успешный вход",
		Data:    token,
		User: models.UserResponse{
			ID:        user.ID,
			FirstName: user.FirstName,
			LastName:  user.LastName,
			Email:     user.Email,
			Role:      user.Role,
			Name:      user.Name,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func GetCurrentUser(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value(models.UserContextKey).(string)

	user, err := repository.GetUserByID(r.Context(), userID)
	if err != nil {
		http.Error(w, "user doesnt exists", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Message: "User is authenticated",
		User: models.UserResponse{
			ID:        user.ID,
			FirstName: user.FirstName,
			LastName:  user.LastName,
			Email:     user.Email,
			Role:      user.Role,
			Name:      user.Name,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func UpdateProfile(w http.ResponseWriter, r *http.Request) {
	var req models.UpdateProfileRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	response := models.Response{
		Success: true,
		Message: "Profile updated successfully",
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func ChangePassword(w http.ResponseWriter, r *http.Request) {
	var req models.ChangePasswordRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}
	userID := r.Context().Value(models.UserContextKey).(string)

	var hashedPassword string
	err = database.DB.QueryRow("SELECT password FROM users WHERE id = $1", userID).Scan(&hashedPassword)
	if err != nil {
		http.Error(w, "User not found", http.StatusNotFound)
		return
	}

	err = bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(req.CurrentPassword))
	if err != nil {
		http.Error(w, "Current password is incorrect", http.StatusUnauthorized)
		return
	}

	newHashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.NewPassword), bcrypt.DefaultCost)
	if err != nil {
		http.Error(w, "Failed to hash password", http.StatusInternalServerError)
		return
	}

	err = repository.UpdatePassword(r.Context(), userID, string(newHashedPassword))
	if err != nil {
		http.Error(w, "Failed to update password", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Message: "Password changed successfully",
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func ForgotPassword(w http.ResponseWriter, r *http.Request) {
	var req models.ForgotPasswordRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	response := models.Response{
		Success: true,
		Message: "Password reset instructions sent",
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func ResetPassword(w http.ResponseWriter, r *http.Request) {
	var req models.ResetPasswordRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	response := models.Response{
		Success: true,
		Message: "Password has been reset successfully",
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}
